### PR TITLE
Documentation for Query DSL #365

### PR DESCRIPTION
**Pull request by vbradnitski**

### CODE REVIEW

File: docs/storage/dsl.adoc [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+
+*Example requests:*
+
+.Matches if any of query's words found in listed properties and boosts the match in '_name' field
+[source,json]
+----
+{
+  "fulltext": {
+    "fields": [
+      "_name^3",
```
Comment (rymsha) [inline docs/storage/dsl.adoc:281]: I don't see ^ documented anywhere

File: docs/storage/dsl.adoc [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+    "field": "displayName",
+    "query": "~apple pork+pie*"
+  }
+}
+----
+
+==== ngram
+
+An n-gram is a sequence of n letters from a term.
+During ngram indexing, the term "foxy" is also indexed as: "f", "fo", and "fox".
```
Comment (rymsha) [inline docs/storage/dsl.adoc:306]: This is actually called "edge n-gram".

**Review by rymsha — CHANGES_REQUESTED** [review]
I think JS API need to be slightly updated, too

File: docs/storage/dsl.adoc [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+.Matches if the field has one of exact values in typed index.
+[source,json]
+----
+{
+  "in": {
+    "field": "myDateTime",
+    "type": "dateTime",
+    "values": [
+      "2015-02-26T12:00:00.030Z",
+      "2015-02-26T12:00:00-02:23"
```
Comment (rymsha) [inline docs/storage/dsl.adoc:78]:
```suggestion
      "2015-02-26T12:00:00-02:23:00Z"
```

File: docs/storage/dsl.adoc [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+.Matches if "myDateTime" field contains any of the <<dateTime,dateTime>> indexed values.
+[source,json]
+----
+{
+  "in": {
+    "field": "myDateTime",
+    "type": "dateTime",
+    "values": [
+      "2015-02-26T12:00:00.030Z",
+      "2015-02-26T12:00:00-02:23"
```
Comment (rymsha) [inline docs/storage/dsl.adoc:114]:
```suggestion
      "2015-02-26T12:00:00-02:23:00Z"
```

File: docs/storage/dsl.adoc [hunk 5]
```diff
@@ -0,0 +1,678 @@
...
+* ``-`` negates a single token.
+* ``*`` at the end of a term signifies a prefix query.
+* ``(`` and ``)`` signify precedence.
+* ``"`` and ``"`` wraps a number of tokens to signify a phrase for searching
+* ``~N`` after a word signifies edit distance (fuzziness) with a number representing https://en.wikipedia.org/wiki/Levenshtein_distance[Levenshtein distance].
+* ``~N`` after a phrase signifies slop amount (how far apart terms in phrase are allowed)
+
+To use one of these characters literally, escape it with a preceding backslash (\).
+
+NOTE: You can boost - thus increasing or decreasing hit-score pr field basis. By providing more than one field to the query by appending a weight-factor: ^N
```
Comment (rymsha) [inline docs/storage/dsl.adoc:263]:
```suggestion
NOTE: You can boost - thus increasing or decreasing hit-score per field basis. By providing more than one field to the query by appending a weight-factor: ^N
```

File: docs/storage/dsl.adoc [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+Date expression makes a search for fields indexed as 'datetime' (`Instant`, `LocalDate`, `LocalDateTime` <<indexing#value_types, value types>>). Value will be converted to zulu time with zero time and UTC timezone.
+
+==== dateTime
+
+DateTime expression makes a search for fields indexed as 'datetime' (`Instant`, `LocalDate`, `LocalDateTime` <<indexing#value_types, value types>>).
+
+==== localDateTime
+
+LocalDateTime expression makes a search for fields indexed as 'datetime' ( `Instant`, `LocalDate`, `LocalDateTime` <<indexing#value_types, value types>>). UTC timezone is used for search.
+
```
Comment (rymsha) [inline docs/storage/dsl.adoc:39]: `time` is missing

**Review by rymsha — CHANGES_REQUESTED** [review]

**Review by rymsha — APPROVED** [review]

**Review by alansemenov — CHANGES_REQUESTED** [review]

**Comment by alansemenov** [comment]
I've changed a few *Example requests:*, but they should be replaced everywhere in the document. It should be either "_Sample requests_" or "_Request examples_".

**Review by alansemenov — APPROVED** [review]

**Pull request merged**